azure/services/aso/aso.go

```diff
@@ -124,33 +133,25 @@ func (s *Service) CreateOrUpdateResource(ctx context.Context, spec azure.ASOReso
 
 			if readyErr != nil {
 				if conds[i].Severity == conditions.ConditionSeverityError {
-					return nil, azure.WithTerminalError(readyErr)
+					return zero, azure.WithTerminalError(readyErr)
 				}
-				return nil, azure.WithTransientError(readyErr, requeueInterval)
+				return zero, azure.WithTransientError(readyErr, requeueInterval)
 			}
 		}
 	}
 
 	// Construct parameters using the resource spec and information from the existing resource, if there is one.
-	parameters, err := spec.Parameters(ctx, deepCopyOrNil(existing))
+	parameters, err := spec.Parameters(ctx, existing.DeepCopy())
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to get desired parameters for resource %s/%s (service: %s)", resourceNamespace, resourceName, serviceName)
-	}
-	// a nil result here is a special case for compatibility with the old
-	// SDK-driven service implementations.
```

Comment:
Because of the inability to compare `T` to `nil`, I had to walk back this workaround. Now implementations of `Parameters` must return the `existing` input object unmodified if no updates are required instead of returning `nil`.